contracts/staking/GoodCompoundStaking.sol

```diff
@@ -4,91 +4,125 @@ pragma solidity >=0.7.0;
+	 * @param _tokenSymbol Symbol of the staking token which will be provided to staker for their staking share
+	 * @param _tokenSymbol Determines blocks to pass for 1x Multiplier
+	 */
+	constructor(
+		address _token,
+		address _iToken,
+		uint256 _blockInterval,
+		NameService _ns,
+		string memory _tokenName,
+		string memory _tokenSymbol,
+		uint64 _maxRewardThreshold
+	)
+		SimpleStaking(
+			_token,
+			_iToken,
+			_blockInterval,
+			_ns,
+			_tokenName,
+			_tokenSymbol,
+			_maxRewardThreshold
+		)
+	{}
+
+	/**
+	 * @dev stake some DAI
+	 * @param _amount of dai to stake
+	 */
+	function mintInterestToken(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		uint256 res = cToken.mint(_amount);
+
+		if (
+			res > 0
+		) //cDAI returns >0 if error happened while minting. make sure no errors, if error return DAI funds
+		{
+			require(res == 0, "Minting cDai failed, funds returned");
+		}
+	}
+
+	/**
+	 * @dev redeem DAI from compound
+	 * @param _amount of dai to redeem
+	 */
+	function redeem(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		require(cToken.redeemUnderlying(_amount) == 0, "Failed to redeem cDai");
+	}
+
+	function redeemUnderlying(uint256 _amount)
+		internal
+		override
+		returns (address, uint256)
+	{
+		uint256 tokenBalance = token.balanceOf(address(this));
+		cERC20 cToken = cERC20(address(iToken));
+		require(cToken.redeem(_amount) == 0, "Failed to redeem cDai");
+		uint256 redeemedAmount = token.balanceOf(address(this)) - tokenBalance;
+		return (address(token), redeemedAmount);
+	}
+
+	/**
+	 * @dev returns Dai to cDai Exchange rate.
+	 */
+	function exchangeRate() internal view override returns (uint256) {
+		cERC20 cToken = cERC20(address(iToken));
+		return cToken.exchangeRateStored();
+	}
+
+	/**
+	 * @dev returns decimals of token.
+	 */
+	function tokenDecimal() internal view override returns (uint256) {
+		ERC20 token = ERC20(address(token));
+		return uint256(token.decimals());
+	}
+
+	/**
+	 * @dev returns decimals of interest token.
+	 */
+	function iTokenDecimal() internal view override returns (uint256) {
+		ERC20 cToken = ERC20(address(iToken));
+		return uint256(cToken.decimals());
+	}
+
+	/**
+	 @dev function calculate DAI price in USD 
+	 @dev _amount Amount of DAI to calculate worth of it
+	 @return Returns worth of DAIs in USD
+	 */
+	function getTokenPriceInUSD(uint256 _amount)
+		internal
+		view
+		override
+		returns (uint256)
+	{
+		AggregatorV3Interface daiPriceOracle =
+			AggregatorV3Interface(nameService.getAddress("DAI_USD_ORACLE"));
+		(, int256 daiPriceinUSD, , , ) = daiPriceOracle.latestRoundData();
+		return (uint256(daiPriceinUSD) * _amount) / 1e18; // daiPriceinUSD in 8 decimals and _amount is in 18 decimals so we divide it 1e18 at the end to reduce 8 decimals back
+	}
+
+	function getGasCostForInterestTransfer()
+		external
+		view
+		override
+		returns (uint256)
+	{
+		return uint256(67917);
```

Comment:
how do you know? we need to examine real cases we have just mocks

Reply:
Yes maybe should be settable by DAO?

Reply:
yes